File: catalog/skills/doc-gen/impl/doc_gen.py
"""
Doc generation skill.

Constructs an offer packet from a candidate profile and optional compensation
context with JSON Schema validation. This Phase 2 implementation already uses
the async skill signature and accepts an optional MCP runtime, but it currently
relies on local data transformation only.
"""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

import jsonschema
import yaml

from magsag.mcp import MCPRuntime

TEMPLATE_QUERY = """
SELECT
    summary_template,
    talking_points_template,
    default_warnings,
    provenance_inputs,
    provenance_schemas
FROM offer_templates
WHERE slug = $1 OR slug = 'default'
ORDER BY CASE WHEN slug = $1 THEN 0 ELSE 1 END
LIMIT 1
"""

ROOT = Path(__file__).resolve().parents[4]  # Point to repo root
INPUT_CONTRACT = ROOT / "catalog" / "contracts" / "candidate_profile.schema.json"
OUTPUT_CONTRACT = ROOT / "catalog" / "contracts" / "offer_packet.schema.json"


def _load_schema(path: Path) -> dict[str, Any]:
    data = yaml.safe_load(path.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError(f"Schema at {path} must be a JSON object")
    return data


def _relative_path(path: Path) -> str:
    try:
        return str(path.resolve().relative_to(ROOT))
    except ValueError:  # pragma: no cover - defensive
        return str(path)


INPUT_SCHEMA = _load_schema(INPUT_CONTRACT)
OUTPUT_SCHEMA = _load_schema(OUTPUT_CONTRACT)


def _validate(payload: dict[str, Any], schema: dict[str, Any], name: str) -> None:
    try:
        jsonschema.validate(payload, schema)
    except jsonschema.ValidationError as exc:  # pragma: no cover - defensive
        raise ValueError(f"{name} schema validation failed: {exc.message}") from exc


def _normalized_candidate(payload: dict[str, Any]) -> dict[str, Any]:
    identifier = str(payload.get("id", "")).strip()
    if not identifier:
        raise ValueError("Candidate profile requires an 'id' field")

    candidate = {
        "id": identifier,
        "name": payload.get("name") or "Unknown Candidate",
        "role": payload.get("role") or payload.get("title") or "Unknown Role",
        "level": payload.get("level") or payload.get("seniority"),
        "location": payload.get("location"),
    }
    return candidate


def _compensation_section(payload: dict[str, Any]) -> dict[str, Any]:
    raw_band = payload.get("salary_band")
    band = raw_band if isinstance(raw_band, dict) else {}
    base = payload.get("base_salary") or band.get("base") or band.get("min")
    maximum = payload.get("max_salary") or band.get("max")
    currency = band.get("currency") or "USD"
    components: dict[str, Any] = {}
    if base is not None:
        components["base"] = {"amount": base, "currency": currency}
    if maximum is not None:
        components["ceiling"] = {"amount": maximum, "currency": currency}

    variable = payload.get("variable_comp")
    if variable:
        components["variable"] = variable

    equity = payload.get("equity")
    if equity:
        components["equity"] = equity

    recommendation = payload.get("compensation_recommendation")
    if recommendation:
        components["recommendation"] = recommendation

    source = payload.get("salary_band_source") or band.get("source")
    return {
        "components": components,
        "source": source,
    }


def _build_narrative(candidate: dict[str, Any], compensation: dict[str, Any]) -> dict[str, str]:
    name = candidate.get("name") or candidate["id"]
    role = candidate.get("role") or "the target role"
    base_component = compensation["components"].get("base")
    if base_component:
        comp_phrase = f"a base salary of {base_component['amount']} {base_component['currency']}"
    else:
        comp_phrase = "a competitive base salary aligned with market data"

    highlights = [
        f"Recommend {comp_phrase} for {name} ({role}).",
        "Total compensation aligns with market benchmarks and advisor guidance.",
    ]

    if compensation.get("source"):
        highlights.append(f"Compensation source: {compensation['source']}.")

    return {
        "summary": " ".join(highlights),
        "talking_points": "\n".join(highlights),
    }


def _collect_warnings(payload: dict[str, Any], compensation: dict[str, Any]) -> list[str]:
    warnings: list[str] = []
    if not compensation["components"]:
        warnings.append("Salary band information is missing; confirm compensation details.")
    if payload.get("advisor_notes") is None:
        warnings.append("Advisor notes not supplied.")
    if payload.get("salary_band") is None:
        warnings.append("Salary band lookup result not attached.")
    return warnings


def _template_slug(candidate: dict[str, Any]) -> str:
    role = (candidate.get("role") or "").strip().lower() or "unknown-role"
    level = (candidate.get("level") or "").strip().lower() or "unknown-level"
    return f"{role}:{level}"


def _render_template(template: str, context: dict[str, Any]) -> str:
    class _Safe(dict):
        def __missing__(self, key: str) -> str:  # pragma: no cover - defensive
            return f"{{{key}}}"

    return template.format_map(_Safe(context))


async def _load_offer_template(slug: str, mcp: MCPRuntime) -> dict[str, Any]:
    result = await mcp.query_postgres(
        server_id="pg-readonly",
        sql=TEMPLATE_QUERY,
        params=[slug],
    )

    if not result.success:
        raise RuntimeError(f"Failed to load offer template via MCP: {result.error}")

    rows = (result.output or {}).get("rows", []) if isinstance(result.output, dict) else []
    if not rows:
        raise RuntimeError(f"No offer template found for slug '{slug}' (including default).")

    template_row = rows[0]
    summary_template = template_row.get("summary_template")
    talking_points_template = template_row.get("talking_points_template")
    default_warnings = template_row.get("default_warnings") or []

    if not isinstance(summary_template, str) or not summary_template.strip():
        raise RuntimeError(
            f"Offer template '{slug}' is missing a summary_template string in the database."
        )

    if isinstance(talking_points_template, str):
        talking_points = [line.strip() for line in talking_points_template.split("\n") if line.strip()]
    elif isinstance(talking_points_template, list):
        talking_points = [str(item) for item in talking_points_template]
    else:
        talking_points = []

    if not isinstance(default_warnings, list):
        default_warnings = [str(default_warnings)]

    provenance_inputs = template_row.get("provenance_inputs") or []
    if not isinstance(provenance_inputs, list):
        provenance_inputs = [str(provenance_inputs)]

    provenance_schemas = template_row.get("provenance_schemas") or {}
    if not isinstance(provenance_schemas, dict):
        provenance_schemas = {}

    return {
        "summary_template": summary_template,
        "talking_points": talking_points,
        "default_warnings": [str(item) for item in default_warnings],
        "provenance_inputs": [str(item) for item in provenance_inputs],
        "provenance_schemas": provenance_schemas,
    }


async def run(
    payload: dict[str, Any],
    *,
    mcp: Optional[MCPRuntime] = None,
) -> dict[str, Any]:
    """
    Generate a structured offer packet from a candidate profile.

    CURRENT IMPLEMENTATION: Pure local data transformation with optional MCP runtime
    parameter (unused). Future phases may leverage MCP for template retrieval.

    Args:
        payload: Candidate profile data matching the candidate_profile contract.
        mcp: Optional MCP runtime. Required for Phase 3 to load governed templates.

    Returns:
        Offer packet payload satisfying the offer_packet contract.

    NOTE: Offer templates and baseline warnings are stored in the governed
    PostgreSQL catalog. Missing MCP runtime now results in a runtime error.
    """
    _validate(payload, INPUT_SCHEMA, "candidate_profile")

    if mcp is None:
        raise RuntimeError(
            "doc-gen requires an MCP runtime with access to the 'pg-readonly' server."
        )

    candidate = _normalized_candidate(payload)
    compensation = _compensation_section(payload)
    template_slug = payload.get("template_slug") or _template_slug(candidate)
    offer_template = await _load_offer_template(template_slug, mcp)

    base_component = compensation["components"].get("base", {})
    base_phrase = "market-aligned base salary"
    base_amount = base_component.get("amount")
    base_currency = base_component.get("currency", "USD")
    if base_amount is not None:
        base_phrase = f"{base_amount} {base_currency}"

    context = {
        "candidate_name": candidate.get("name") or candidate["id"],
        "candidate_role": candidate.get("role") or "Unknown Role",
        "candidate_level": candidate.get("level") or "Unknown Level",
        "location": candidate.get("location") or "Unknown Location",
        "base_salary_amount": base_amount if base_amount is not None else "",
        "base_salary_currency": base_currency,
        "base_salary_phrase": base_phrase,
        "base_salary": base_amount if base_amount is not None else base_phrase,
        "advisor_notes": payload.get("advisor_notes", ""),
    }

    narrative = {
        "summary": _render_template(offer_template["summary_template"], context),
        "talking_points": "\n".join(
            [_render_template(point, context) for point in offer_template["talking_points"]]
        )
        or _build_narrative(candidate, compensation)["talking_points"],
    }

    warnings = list(
        dict.fromkeys(offer_template["default_warnings"] + _collect_warnings(payload, compensation))
    )

    offer_id = payload.get("offer_id") or f"offer-{candidate['id']}"

    salary_band_payload = payload.get("salary_band") or {}
    band_currency = salary_band_payload.get("currency") or base_currency
    band_min = (
        salary_band_payload.get("min")
        or salary_band_payload.get("base")
        or base_amount
        or 0
    )
    band_max = salary_band_payload.get("max") or salary_band_payload.get("ceiling") or band_min

    offer_block: dict[str, Any] = {
        "role": candidate.get("role") or "Unknown Role",
        "base_salary": {
            "currency": base_currency,
            "amount": float(base_amount) if base_amount is not None else float(band_min),
        },
        "band": {
            "currency": band_currency,
            "min": float(band_min),
            "max": float(band_max if band_max is not None else band_min),
        },
    }

    resolved_band_source = compensation.get("source") or salary_band_payload.get("source")
    if resolved_band_source:
        offer_block["band"]["source"] = resolved_band_source

    sign_on_bonus = payload.get("sign_on_bonus")
    if isinstance(sign_on_bonus, dict):
        offer_block["sign_on_bonus"] = sign_on_bonus

    equity = payload.get("equity")
    if isinstance(equity, dict):
        offer_block["equity"] = equity

    provenance_schemas = (
        {**offer_template["provenance_schemas"]}
        if offer_template["provenance_schemas"]
        else {}
    )
    provenance_schemas.setdefault("input", _relative_path(INPUT_CONTRACT))
    provenance_schemas.setdefault("output", _relative_path(OUTPUT_CONTRACT))

    timestamp = datetime.now(timezone.utc).isoformat()

    result: dict[str, Any] = {
        "offer_id": str(offer_id),
        "generated_at": timestamp,
        "offer": offer_block,
        "candidate": candidate,
        "compensation": compensation,
        "narrative": narrative,
        "warnings": warnings,
        "provenance": {
            "schemas": provenance_schemas,
            "inputs": offer_template["provenance_inputs"] or ["candidate_profile"],
        },
        "metadata": {
            "generated_by": "skill.doc-gen",
            "timestamp": timestamp,
            "version": "0.1.0",
        },
    }

    _validate(result, OUTPUT_SCHEMA, "offer_packet")
    return result

File: packages/catalog/src/skills/doc-gen.ts
import { createRequire } from 'node:module';
import type { ValidateFunction } from 'ajv';
import { SkillContext, McpRuntime, McpToolResult } from '../shared/types.js';
import candidateProfileSchemaJson from '../../../../catalog/contracts/candidate_profile.schema.json' with { type: 'json' };
import offerPacketSchemaJson from '../../../../catalog/contracts/offer_packet.schema.json' with { type: 'json' };

const moduleRequire = createRequire(import.meta.url);

const TEMPLATE_QUERY = `
SELECT
    summary_template,
    talking_points_template,
    default_warnings,
    provenance_inputs,
    provenance_schemas
FROM offer_templates
WHERE slug = $1 OR slug = 'default'
ORDER BY CASE WHEN slug = $1 THEN 0 ELSE 1 END
LIMIT 1
`;

const INPUT_SCHEMA_PATH = 'catalog/contracts/candidate_profile.schema.json';
const OUTPUT_SCHEMA_PATH = 'catalog/contracts/offer_packet.schema.json';

const AjvModule = moduleRequire('ajv/dist/2020.js') as { default?: typeof import('ajv').default };
const AjvFactory = (AjvModule.default ?? AjvModule) as typeof import('ajv').default;
const addFormatsModule = moduleRequire('ajv-formats') as { default?: (ajvInstance: unknown) => void };
const addFormats = (addFormatsModule.default ?? addFormatsModule) as (ajvInstance: unknown) => void;

const ajv = new AjvFactory({ strict: false, allErrors: true });
addFormats(ajv);
const validatorCache = new Map<string, ValidateFunction>();
const schemaCache: Record<string, Record<string, unknown>> = {
  [INPUT_SCHEMA_PATH]: candidateProfileSchemaJson as Record<string, unknown>,
  [OUTPUT_SCHEMA_PATH]: offerPacketSchemaJson as Record<string, unknown>
};

const compileSchema = (pathKey: string): ValidateFunction => {
  const memoized = validatorCache.get(pathKey);
  if (memoized) {
    return memoized;
  }

  const schema = schemaCache[pathKey];
  if (!schema) {
    throw new Error(`Unknown schema requested: ${pathKey}`);
  }

  const validator = ajv.compile(schema);
  validatorCache.set(pathKey, validator);
  return validator;
};

const validateSchema = (pathKey: string, data: unknown, label: string): void => {
  const validator = compileSchema(pathKey);
  if (validator(data)) {
    return;
  }
  const errors = validator.errors ?? [];
  const details =
    errors.length > 0
      ? errors
          .map((item) => {
            const instancePath = item.instancePath && item.instancePath.length > 0 ? item.instancePath : '/';
            const message = item.message ?? 'is invalid';
            return `${instancePath} ${message}`.trim();
          })
          .join('; ')
      : 'validation failed';
  throw new Error(`${label} validation failed: ${details}`);
};

type OfferTemplate = {
  summaryTemplate: string;
  talkingPoints: string[];
  defaultWarnings: string[];
  provenanceInputs: string[];
  provenanceSchemas: Record<string, unknown>;
};

const ensureMcp = (context: SkillContext): McpRuntime => {
  if (!context.mcp) {
    throw new Error("doc-gen requires an MCP runtime with access to the 'pg-readonly' server.");
  }
  return context.mcp;
};

const asRecord = (value: unknown): Record<string, unknown> | null =>
  value !== null && typeof value === 'object' ? (value as Record<string, unknown>) : null;

const toFiniteNumber = (value: unknown, field: string): number | undefined => {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) {
      throw new Error(`${field} must be a finite number`);
    }
    return value;
  }
  if (typeof value === 'string') {
    const trimmed = value.trim();
    if (trimmed.length === 0) {
      return undefined;
    }
    const parsed = Number(trimmed);
    if (!Number.isFinite(parsed)) {
      throw new Error(`${field} must be a finite number`);
    }
    return parsed;
  }
  throw new Error(`${field} must be a finite number`);
};

const asNonEmptyString = (value: unknown): string | undefined => {
  if (typeof value !== 'string') {
    return undefined;
  }
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
};

const normalizeCandidate = (payload: Record<string, unknown>): Record<string, unknown> => {
  const identifier = String(payload.id ?? '').trim();
  if (!identifier) {
    throw new Error("Candidate profile requires an 'id' field");
  }
  return {
    id: identifier,
    name: asNonEmptyString(payload.name) ?? 'Unknown Candidate',
    role: asNonEmptyString(payload.role) ?? asNonEmptyString(payload.title) ?? 'Unknown Role',
    level: asNonEmptyString(payload.level) ?? asNonEmptyString(payload.seniority),
    location: asNonEmptyString(payload.location)
  };
};

const compensationSection = (payload: Record<string, unknown>): Record<string, unknown> => {
  const rawBand = asRecord(payload.salary_band);
  const band = rawBand ?? {};
  const currency = band.currency ?? 'USD';

  const baseAmount =
    toFiniteNumber(payload.base_salary, 'payload.base_salary') ??
    toFiniteNumber(band.base, 'salary_band.base') ??
    toFiniteNumber(band.min, 'salary_band.min');

  const ceilingAmount =
    toFiniteNumber(payload.max_salary, 'payload.max_salary') ??
    toFiniteNumber(band.max, 'salary_band.max') ??
    toFiniteNumber(band.ceiling, 'salary_band.ceiling');

  const components: Record<string, unknown> = {};
  if (baseAmount !== undefined) {
    components.base = { amount: baseAmount, currency };
  }
  if (ceilingAmount !== undefined) {
    components.ceiling = { amount: ceilingAmount, currency };
  }
  if (payload.variable_comp) {
    components.variable = payload.variable_comp;
  }
  if (payload.equity) {
    components.equity = payload.equity;
  }
  if (payload.compensation_recommendation) {
    components.recommendation = payload.compensation_recommendation;
  }

  return {
    components,
    source: payload.salary_band_source ?? band.source
  };
};

const collectWarnings = (
  payload: Record<string, unknown>,
  compensation: Record<string, unknown>
): string[] => {
  const warnings: string[] = [];
  const components = asRecord(compensation.components) ?? {};
  if (Object.keys(components).length === 0) {
    warnings.push('Salary band information is missing; confirm compensation details.');
  }
  if (!payload.advisor_notes) {
    warnings.push('Advisor notes not supplied.');
  }
  const salaryBandRecord = asRecord(payload.salary_band);
  if (!salaryBandRecord || Object.keys(salaryBandRecord).length === 0) {
    warnings.push('Salary band lookup result not attached.');
  }
  return warnings;
};

const templateSlug = (candidate: Record<string, unknown>): string => {
  const role = String(candidate.role ?? '').trim().toLowerCase() || 'unknown-role';
  const level = String(candidate.level ?? '').trim().toLowerCase() || 'unknown-level';
  return `${role}:${level}`;
};

const renderTemplate = (template: string, context: Record<string, unknown>): string => {
  return template.replace(/\{([^}]+)\}/g, (_match, key) => {
    const value = context[key as string];
    return value === undefined ? `{${key}}` : String(value);
  });
};

const ensureToolResult = (result: McpToolResult | undefined, slug: string): OfferTemplate => {
  if (!result || !result.success) {
    const errorMessage = result?.error ?? 'unknown error';
    throw new Error(`Failed to load offer template via MCP: ${errorMessage}`);
  }
  const payload = asRecord(result.output);
  const rows = Array.isArray(payload?.rows) ? payload?.rows : [];
  const firstRow = asRecord(rows?.[0]);
  if (!firstRow) {
    throw new Error(`No offer template found for slug '${slug}' (including default).`);
  }

  const summaryTemplate = typeof firstRow.summary_template === 'string' ? firstRow.summary_template : '';
  if (!summaryTemplate.trim()) {
    throw new Error(
      `Offer template '${slug}' is missing a summary_template string in the database.`
    );
  }

  const talkingPointsRaw = firstRow.talking_points_template;
  let talkingPoints: string[] = [];
  if (typeof talkingPointsRaw === 'string') {
    talkingPoints = talkingPointsRaw
      .split('\n')
      .map((item) => item.trim())
      .filter((item) => item.length > 0);
  } else if (Array.isArray(talkingPointsRaw)) {
    talkingPoints = talkingPointsRaw.map((item) => String(item));
  }

  const defaultWarnings = Array.isArray(firstRow.default_warnings)
    ? firstRow.default_warnings.map((item) => String(item))
    : firstRow.default_warnings
    ? [String(firstRow.default_warnings)]
    : [];

  const provenanceInputs = Array.isArray(firstRow.provenance_inputs)
    ? firstRow.provenance_inputs.map((item) => String(item))
    : firstRow.provenance_inputs
    ? [String(firstRow.provenance_inputs)]
    : [];

  const provenanceSchemas =
    asRecord(firstRow.provenance_schemas) ?? (firstRow.provenance_schemas ? { raw: firstRow.provenance_schemas } : {});

  return {
    summaryTemplate,
    talkingPoints,
    defaultWarnings,
    provenanceInputs,
    provenanceSchemas
  };
};

const loadOfferTemplate = async (
  slug: string,
  runtime: McpRuntime
): Promise<OfferTemplate> => {
  const result = await runtime.queryPostgres?.({
    serverId: 'pg-readonly',
    sql: TEMPLATE_QUERY,
    params: [slug]
  });
  return ensureToolResult(result, slug);
};

const buildNarrativeFallback = (
  candidate: Record<string, unknown>,
  compensation: Record<string, unknown>
): { summary: string; talking_points: string } => {
  const components = asRecord(compensation.components) ?? {};
  const baseComponent = asRecord(components.base);
  const candidateName = candidate.name ?? candidate.id ?? 'the candidate';
  const role = candidate.role ?? 'the target role';
  const basePhrase =
    baseComponent && baseComponent.amount !== undefined
      ? `${baseComponent.amount} ${baseComponent.currency ?? 'USD'}`
      : 'a competitive base salary aligned with market data';
  const lines = [
    `Recommend ${basePhrase} for ${candidateName} (${role}).`,
    'Total compensation aligns with market benchmarks and advisor guidance.'
  ];
  if (compensation.source) {
    lines.push(`Compensation source: ${compensation.source}.`);
  }
  return {
    summary: lines.join(' '),
    talking_points: lines.join('\n')
  };
};

export const run = async (
  payload: Record<string, unknown>,
  context: SkillContext = {}
): Promise<Record<string, unknown>> => {
  validateSchema(INPUT_SCHEMA_PATH, payload, 'doc-gen input');
  const mcp = ensureMcp(context);
  const candidate = normalizeCandidate(payload);
  const compensation = compensationSection(payload);
  const slug =
    typeof payload.template_slug === 'string' && payload.template_slug.length > 0
      ? payload.template_slug
      : templateSlug(candidate);

  const template = await loadOfferTemplate(slug, mcp);

  const baseComponent = asRecord(compensation.components)?.base;
  const baseRecord = asRecord(baseComponent);
  const baseAmount = toFiniteNumber(baseRecord?.amount, 'compensation.components.base.amount');
  const baseCurrency = baseRecord?.currency ?? 'USD';

  const contextMap: Record<string, unknown> = {
    candidate_name: candidate.name ?? candidate.id,
    candidate_role: candidate.role ?? 'Unknown Role',
    candidate_level: candidate.level ?? 'Unknown Level',
    location: candidate.location ?? 'Unknown Location',
    base_salary_amount: baseAmount ?? '',
    base_salary_currency: baseCurrency,
    base_salary_phrase:
      baseAmount !== undefined ? `${baseAmount} ${baseCurrency}` : 'market-aligned base salary',
    base_salary: baseAmount ?? 'market-aligned base salary',
    advisor_notes: payload.advisor_notes ?? ''
  };

  const summary = renderTemplate(template.summaryTemplate, contextMap);
  const talkingPointsText = template.talkingPoints
    .map((point) => renderTemplate(point, contextMap))
    .join('\n');
  const fallbackNarrative = buildNarrativeFallback(candidate, compensation);

  const narrative = {
    summary,
    talking_points: talkingPointsText || fallbackNarrative.talking_points
  };

  const warnings = Array.from(
    new Set([...template.defaultWarnings, ...collectWarnings(payload, compensation)])
  );

  const offerId =
    typeof payload.offer_id === 'string' && payload.offer_id.length > 0
      ? payload.offer_id
      : `offer-${candidate.id}`;

  const salaryBand = asRecord(payload.salary_band) ?? {};
  const bandCurrency = salaryBand.currency ?? baseCurrency;
  const salaryBandMin =
    toFiniteNumber(salaryBand.min, 'salary_band.min') ??
    toFiniteNumber(salaryBand.base, 'salary_band.base');
  const resolvedBandMin = salaryBandMin ?? baseAmount ?? 0;
  if (!Number.isFinite(resolvedBandMin)) {
    throw new Error('Resolved salary band minimum must be a finite number');
  }

  const salaryBandMax =
    toFiniteNumber(salaryBand.max, 'salary_band.max') ??
    toFiniteNumber(salaryBand.ceiling, 'salary_band.ceiling');
  const resolvedBandMax = salaryBandMax ?? resolvedBandMin;
  if (!Number.isFinite(resolvedBandMax)) {
    throw new Error('Resolved salary band maximum must be a finite number');
  }

  const resolvedBaseAmount = baseAmount ?? resolvedBandMin;

  const offerBlock: Record<string, unknown> = {
    role: candidate.role ?? 'Unknown Role',
    base_salary: {
      currency: baseCurrency,
      amount: resolvedBaseAmount
    },
    band: {
      currency: bandCurrency,
      min: resolvedBandMin,
      max: resolvedBandMax
    }
  };

  const bandSource = compensation.source ?? salaryBand.source;
  if (bandSource) {
    (offerBlock.band as Record<string, unknown>).source = bandSource;
  }

  if (asRecord(payload.sign_on_bonus)) {
    offerBlock.sign_on_bonus = payload.sign_on_bonus;
  }
  if (asRecord(payload.equity)) {
    offerBlock.equity = payload.equity;
  }

  const provenanceSchemas = { ...template.provenanceSchemas };
  provenanceSchemas.input ??= INPUT_SCHEMA_PATH;
  provenanceSchemas.output ??= OUTPUT_SCHEMA_PATH;

  const timestamp = new Date().toISOString();

  const result = {
    offer_id: offerId,
    generated_at: timestamp,
    offer: offerBlock,
    candidate,
    compensation,
    narrative,
    warnings,
    provenance: {
      schemas: provenanceSchemas,
      inputs: template.provenanceInputs.length > 0 ? template.provenanceInputs : ['candidate_profile']
    },
    metadata: {
      generated_by: 'skill.doc-gen',
      timestamp,
      version: '0.1.0'
    }
  };

  validateSchema(OUTPUT_SCHEMA_PATH, result, 'doc-gen output');
  return result;
};
